File: src/main/kotlin/schemas/InstallerManifest.kt
package schemas

import kotlinx.serialization.SerialName
import kotlinx.serialization.Serializable

@Serializable
data class InstallerManifest(
    @SerialName("PackageIdentifier") val packageIdentifier: String? = null,
    @SerialName("PackageVersion") val packageVersion: String? = null,
    @SerialName("Channel") val channel: String? = null,
    @SerialName("Commands") val commands: List<String>? = null,
    @SerialName("Protocols") val protocols: List<String>? = null,
    @SerialName("FileExtensions") val fileExtensions: List<String>? = null,
    @SerialName("Installers") val installers: List<Installer> = listOf(),
    @SerialName("ManifestType") val manifestType: String = "",
    @SerialName("ManifestVersion") val manifestVersion: String = ""
) {
    @Serializable
    data class Installer(
        @SerialName("Architecture") val architecture: String? = null,
        @SerialName("InstallerLocale") val installerLocale: String? = null,
        @SerialName("Platform") val platform: String? = null,
        @SerialName("MinimumOSVersion") val minimumOSVersion: String? = null,
        @SerialName("InstallerType") val installerType: String? = null,
        @SerialName("InstallerUrl") val installerUrl: String? = null,
        @SerialName("InstallerSha256") val installerSha256: String? = null,
        @SerialName("SignatureSha256") val signatureSha256: String? = null,
        @SerialName("Scope") val scope: String? = null,
        @SerialName("InstallModes") val installModes: String? = null,
        @SerialName("InstallerSwitches") val installerSwitches: InstallerSwitches = InstallerSwitches(),
        @SerialName("UpgradeBehavior") val upgradeBehavior: String? = null,
        @SerialName("Dependencies") val dependencies: List<Dependency> = listOf(),
        @SerialName("PackageFamilyName") val packageFamilyName: String? = null,
        @SerialName("Capabilities") val capabilities: String? = null,
        @SerialName("RestrictedCapabilities") val restrictedCapabilities: String? = null,
        @SerialName("InstallerAbortsTerminal") val installerAbortsTerminal: String? = null,
        @SerialName("InstallLocationRequired") val installLocationRequired: String? = null,
        @SerialName("RequireExplicitUpgrade") val requireExplicitUpgrade: String? = null,
        @SerialName("ElevationRequirement") val elevationRequirement: String? = null,
        @SerialName("UnsupportedOSArchitectures") val unsupportedOSArchitectures: String? = null,
        @SerialName("Markets") val markets: String? = null,
        @SerialName("ExcludedMarkets") val excludedMarkets: String? = null,
        @SerialName("InstallerSuccessCodes") val installerSuccessCodes: String? = null,
        @SerialName("ExpectedReturnCodes") val expectedReturnCodes: List<ExpectedReturnCode> = listOf(),
        @SerialName("ProductCode") val productCode: String? = null,
        @SerialName("AppsAndFeaturesEntries") val appsAndFeaturesEntries: List<AppsAndFeaturesEntry> = listOf(),
        @SerialName("UnsupportedArguments") val unsupportedArguments: List<UnsupportedArgument> = listOf(),
        @SerialName("DisplayInstallWarnings") val displayInstallWarnings: String? = null,
        @SerialName("ReleaseDate") val releaseDate: String? = null
    ) {
        @Serializable
        data class InstallerSwitches(
            @SerialName("Silent") val silent: String? = null,
            @SerialName("SilentWithProgress") val silentWithProgress: String? = null,
            @SerialName("Interactive") val interactive: String? = null,
            @SerialName("InstallLocation") val installLocation: String? = null,
            @SerialName("Log") val log: String? = null,
            @SerialName("Upgrade") val upgrade: String? = null,
            @SerialName("Custom") val custom: String? = null
        )

        @Serializable
        data class Dependency(
            @SerialName("ExternalDependencies") val externalDependencies: String? = null,
            @SerialName("PackageDependencies") val packageDependencies: String? = null,
            @SerialName("WindowsFeatures") val windowsFeatures: String? = null,
            @SerialName("WindowsLibraries") val windowsLibraries: String? = null
        )

        @Serializable
        data class ExpectedReturnCode(
            @SerialName("ExpectedReturnCode") val expectedReturnCode: String? = null,
            @SerialName("ReturnResponse") val returnResponse: String? = null,
            @SerialName("ReturnResponseUrl") val returnResponseUrl: String? = null
        )

        @Serializable
        data class AppsAndFeaturesEntry(
            @SerialName("DisplayName") val displayName: String? = null,
            @SerialName("DisplayVersion") val displayVersion: String? = null,
            @SerialName("Publisher") val publisher: String? = null,
            @SerialName("ProductCode") val productCode: String? = null,
            @SerialName("UpgradeCode") val upgradeCode: String? = null,
            @SerialName("InstallerType") val installerType: String? = null
        )

        @Serializable
        data class UnsupportedArgument(
            @SerialName("UnsupportedArgument") val unsupportedArgument: String? = null
        )
    }
}

File: src/installer_manifest.rs
use crate::types::command::Command;
use crate::types::custom_switch::CustomSwitch;
use crate::types::file_extension::FileExtension;
use crate::types::installer_success_code::InstallerSuccessCode;
use crate::types::installer_switch::InstallerSwitch;
use crate::types::language_tag::LanguageTag;
use crate::types::manifest_type::ManifestType;
use crate::types::manifest_version::ManifestVersion;
use crate::types::minimum_os_version::MinimumOSVersion;
use crate::types::package_identifier::PackageIdentifier;
use crate::types::package_version::PackageVersion;
use crate::types::protocol::Protocol;
use crate::types::silent_switch::SilentSwitch;
use crate::types::silent_with_progress_switch::SilentWithProgressSwitch;
use derivative::Derivative;
use serde::{Deserialize, Serialize};
use serde_with::skip_serializing_none;
use std::collections::BTreeSet;
use std::num::NonZeroI64;
use strum::{Display, EnumIter, EnumString};
use time::Date;
use url::Url;

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Default)]
#[serde(rename_all = "PascalCase")]
pub struct InstallerManifest {
    pub package_identifier: PackageIdentifier,
    pub package_version: PackageVersion,
    pub channel: Option<String>,
    pub installer_locale: Option<LanguageTag>,
    pub platform: Option<BTreeSet<Platform>>,
    #[serde(rename = "MinimumOSVersion")]
    pub minimum_os_version: Option<MinimumOSVersion>,
    pub installer_type: Option<InstallerType>,
    pub nested_installer_type: Option<InstallerType>,
    pub nested_installer_files: Option<BTreeSet<NestedInstallerFiles>>,
    pub scope: Option<Scope>,
    pub install_modes: Option<BTreeSet<InstallModes>>,
    pub installer_switches: Option<InstallerSwitches>,
    pub installer_success_codes: Option<BTreeSet<InstallerSuccessCode>>,
    pub expected_return_codes: Option<BTreeSet<ExpectedReturnCodes>>,
    pub upgrade_behavior: Option<UpgradeBehavior>,
    pub commands: Option<BTreeSet<Command>>,
    pub protocols: Option<BTreeSet<Protocol>>,
    pub file_extensions: Option<BTreeSet<FileExtension>>,
    pub dependencies: Option<Dependencies>,
    pub package_family_name: Option<String>,
    pub product_code: Option<String>,
    pub capabilities: Option<BTreeSet<String>>,
    pub restricted_capabilities: Option<BTreeSet<String>>,
    pub markets: Option<Markets>,
    pub installer_aborts_terminal: Option<bool>,
    pub release_date: Option<Date>,
    pub installer_location_required: Option<bool>,
    pub require_explicit_upgrade: Option<bool>,
    pub display_install_warnings: Option<bool>,
    #[serde(rename = "UnsupportedOSArchitectures")]
    pub unsupported_os_architectures: Option<BTreeSet<UnsupportedOSArchitectures>>,
    pub unsupported_arguments: Option<BTreeSet<UnsupportedArguments>>,
    pub apps_and_features_entries: Option<BTreeSet<AppsAndFeaturesEntry>>,
    pub elevation_requirement: Option<ElevationRequirement>,
    pub installation_metadata: Option<InstallationMetadata>,
    pub installers: BTreeSet<Installer>,
    pub manifest_type: ManifestType,
    pub manifest_version: ManifestVersion,
}

#[derive(
    Serialize, Deserialize, Clone, Copy, Debug, EnumString, Eq, PartialEq, Hash, Ord, PartialOrd,
)]
pub enum Platform {
    #[serde(rename = "Windows.Desktop")]
    #[strum(serialize = "Windows.Desktop")]
    WindowsDesktop,
    #[serde(rename = "Windows.Universal")]
    #[strum(serialize = "Windows.Universal")]
    WindowsUniversal,
}

#[derive(Serialize, Deserialize, Clone, Copy, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "lowercase")]
pub enum InstallerType {
    Msix,
    Msi,
    Appx,
    Exe,
    Zip,
    Inno,
    Nullsoft,
    Wix,
    Burn,
    Pwa,
    Portable,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "PascalCase")]
pub struct NestedInstallerFiles {
    pub relative_file_path: String,
    pub portable_command_alias: Option<String>,
}

#[derive(Clone, Copy, Debug, Deserialize, Eq, Hash, Ord, PartialEq, PartialOrd, Serialize)]
#[serde(rename_all = "lowercase")]
pub enum Scope {
    User,
    Machine,
}

#[derive(
    Serialize, Deserialize, Clone, Debug, Display, EnumIter, Eq, PartialEq, Hash, Ord, PartialOrd,
)]
#[serde(rename_all = "camelCase")]
pub enum InstallModes {
    Interactive,
    Silent,
    SilentWithProgress,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, Debug, Default, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "PascalCase")]
pub struct InstallerSwitches {
    pub silent: Option<SilentSwitch>,
    pub silent_with_progress: Option<SilentWithProgressSwitch>,
    pub interactive: Option<InstallerSwitch>,
    pub install_location: Option<InstallerSwitch>,
    pub log: Option<InstallerSwitch>,
    pub upgrade: Option<InstallerSwitch>,
    pub custom: Option<CustomSwitch>,
}

impl InstallerSwitches {
    pub fn are_all_none(&self) -> bool {
        matches!(
            (
                &self.silent,
                &self.silent_with_progress,
                &self.interactive,
                &self.install_location,
                &self.log,
                &self.upgrade,
                &self.custom,
            ),
            (None, None, None, None, None, None, None)
        )
    }
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "PascalCase")]
pub struct ExpectedReturnCodes {
    pub installer_return_code: Option<NonZeroI64>,
    pub return_response: ReturnResponse,
    pub return_response_url: Option<Url>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "camelCase")]
pub enum ReturnResponse {
    PackageInUse,
    PackageInUseByApplication,
    InstallInProgress,
    FileInUse,
    MissingDependency,
    DiskFull,
    InsufficientMemory,
    InvalidParameter,
    NoNetwork,
    ContactSupport,
    RebootRequiredToFinish,
    RebootRequiredForInstall,
    RebootInitiated,
    CancelledByUser,
    AlreadyInstalled,
    Downgrade,
    BlockedByPolicy,
    SystemNotSupported,
    Custom,
}

#[derive(
    Serialize,
    Deserialize,
    Clone,
    Copy,
    Debug,
    Display,
    EnumIter,
    Eq,
    PartialEq,
    Hash,
    Ord,
    PartialOrd,
)]
#[serde(rename_all = "camelCase")]
pub enum UpgradeBehavior {
    Install,
    UninstallPrevious,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "PascalCase")]
pub struct Dependencies {
    pub windows_features: Option<BTreeSet<String>>,
    pub windows_libraries: Option<BTreeSet<String>>,
    pub package_dependencies: Option<PackageDependencies>,
    pub external_dependencies: Option<BTreeSet<String>>,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "PascalCase")]
pub struct PackageDependencies {
    pub package_identifier: String,
    pub minimum_version: Option<String>,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "PascalCase")]
pub struct Markets {
    pub allowed_markets: Option<BTreeSet<String>>,
    pub excluded_markets: Option<BTreeSet<String>>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "lowercase")]
pub enum UnsupportedOSArchitectures {
    X86,
    X64,
    Arm,
    Arm64,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "lowercase")]
pub enum UnsupportedArguments {
    Log,
    Location,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, Debug, Default, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "PascalCase")]
pub struct AppsAndFeaturesEntry {
    pub display_name: Option<String>,
    pub publisher: Option<String>,
    pub display_version: Option<String>,
    pub product_code: Option<String>,
    pub upgrade_code: Option<String>,
    pub installer_type: Option<InstallerType>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "camelCase")]
pub enum ElevationRequirement {
    ElevationRequired,
    ElevationProhibited,
    ElevatesSelf,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "PascalCase")]
pub struct InstallationMetadata {
    pub default_install_location: Option<String>,
    pub files: Option<BTreeSet<MetadataFiles>>,
}

#[skip_serializing_none]
#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "PascalCase")]
pub struct MetadataFiles {
    pub relative_file_path: String,
    pub file_sha_256: Option<String>,
    pub file_type: Option<MetadataFileType>,
    pub invocation_parameter: Option<String>,
    pub display_name: Option<String>,
}

#[derive(Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd)]
#[serde(rename_all = "lowercase")]
pub enum MetadataFileType {
    Launch,
    Uninstall,
    Other,
}

#[skip_serializing_none]
#[derive(
    Serialize, Deserialize, Clone, Debug, Eq, PartialEq, Hash, Ord, PartialOrd, Derivative,
)]
#[derivative(Default)]
#[serde(rename_all = "PascalCase")]
pub struct Installer {
    pub installer_locale: Option<LanguageTag>,
    pub platform: Option<BTreeSet<Platform>>,
    #[serde(rename = "MinimumOSVersion")]
    pub minimum_os_version: Option<MinimumOSVersion>,
    pub architecture: Architecture,
    pub installer_type: Option<InstallerType>,
    pub nested_installer_type: Option<InstallerType>,
    pub nested_installer_files: Option<BTreeSet<NestedInstallerFiles>>,
    pub scope: Option<Scope>,
    #[derivative(Default(value = "Url::parse(\"https://www.example.com\").unwrap()"))]
    pub installer_url: Url,
    pub installer_sha_256: String,
    pub signature_sha_256: Option<String>,
    pub install_modes: Option<BTreeSet<InstallModes>>,
    pub installer_switches: Option<InstallerSwitches>,
    pub installer_success_codes: Option<BTreeSet<InstallerSuccessCode>>,
    pub expected_return_codes: Option<BTreeSet<ExpectedReturnCodes>>,
    pub upgrade_behavior: Option<UpgradeBehavior>,
    pub commands: Option<BTreeSet<Command>>,
    pub protocols: Option<BTreeSet<Protocol>>,
    pub file_extensions: Option<BTreeSet<FileExtension>>,
    pub dependencies: Option<Dependencies>,
    pub package_family_name: Option<String>,
    pub product_code: Option<String>,
    pub capabilities: Option<BTreeSet<String>>,
    pub restricted_capabilities: Option<BTreeSet<String>>,
    pub markets: Option<Markets>,
    pub installer_aborts_terminal: Option<bool>,
    pub release_date: Option<Date>,
    pub installer_location_required: Option<bool>,
    pub require_explicit_upgrade: Option<bool>,
    pub display_install_warnings: Option<bool>,
    #[serde(rename = "UnsupportedOSArchitectures")]
    pub unsupported_os_architectures: Option<BTreeSet<UnsupportedOSArchitectures>>,
    pub unsupported_arguments: Option<BTreeSet<UnsupportedArguments>>,
    pub apps_and_features_entries: Option<BTreeSet<AppsAndFeaturesEntry>>,
    pub elevation_requirement: Option<ElevationRequirement>,
    pub installation_metadata: Option<InstallationMetadata>,
}

#[derive(
    Clone,
    Copy,
    Debug,
    Default,
    Deserialize,
    EnumString,
    Eq,
    Hash,
    Ord,
    PartialEq,
    PartialOrd,
    Serialize,
)]
#[serde(rename_all = "lowercase")]
#[strum(serialize_all = "lowercase")]
pub enum Architecture {
    X86,
    X64,
    Arm,
    Arm64,
    #[default]
    Neutral,
}
